### PR TITLE
fix: privateFieldsAsProperties global counter (#15389)

### LINKED ISSUE #15389: [Bug]: Assumption privateFieldsAsProperties w/o transform-runtime plugin can lead to private field name clashes
### 💻

- [X] Would you like to work on a fix?

### How are you using Babel?

babel-loader (webpack)

### Input code

When using the assumption `privateFieldsAsProperties: true` in combination with not using the Babel `transform-runtime` plugin (so that babel-helpers are inlined), it becomes possible to override #-private fields, which is incorrect semantics.

The issue can only be reproduced when using at least two different source files.

Bar.js:
```js
export default class Bar {
  #hidden = "Bar's hidden";
  getHidden() {
    return this.#hidden;
  }
}
```

index.js:
```js
import Bar from "./Bar";

class Foo extends Bar {
  // it should not be possible to override a #private field of a superclass:
  #hidden = "Foo's hidden";
}

console.log(new Foo().getHidden()); // actual: "Foo's hidden", expected: "Bar's hidden"
```

Since Babel's own REPL only supports one input file, here is the full example including a `.babelrc` in  codesandbox:
[codesandbox.io reproducer](https://codesandbox.io/s/babel-privatefieldsasproperties-bug-gqvfsz?file=/.babelrc:84-88)

### Configuration file name

.babelrc

### Configuration

```json
{
  "presets": [
    "env"
  ],
  "assumptions": {
    "privateFieldsAsProperties": true
  }
}
```

### Current and expected behavior

When running the codesandbox example given above, in the output you see `"Foo's hidden"`, so class `Foo` managed to override the #-private field `#hidden` of class `Bar`, which must not be possible.
When you switch off `privateFieldsAsProperties` and run the example again, the correct result `"Foo's hidden"` is shown.
Also when you use the Babel plugin `transform-runtime`, the behavior is correct.

### Environment

See [codesandbox example](https://codesandbox.io/s/babel-privatefieldsasproperties-bug-gqvfsz?file=/package.json):
- Babel 7.2.0
- Node 16

### Possible solution

The reason that both #-private fields share the same name `__private_0_hidden` is that the helper function [`classPrivateFieldLooseKey`](https://github.com/babel/babel/blob/main/packages/babel-helpers/src/helpers.ts#L997-L1002) tries to define and initialize a global counter variable `id` that is increased for each generated property name, to ensure unique private property names. However, when helper functions are inlined (no `transform-runtime` plugin), each bundled file comes with a _copy_ of this helper code, each defining its own `var id = 0;`. Thus, in both bundled files, the private member counter starts with `0`, ending up in the same private field slot name in both the superclass and the subclass, which must not happen.

One approach to solve this could be to use a `Symbol` instead of a generated string slot name, but I don't know if Babel helper code is allowed to use language features that may not be available in older target environments. I tested this approach and it works perfectly (given that `Symbol` is supported).
I guess helper code is not compiled, so [using `Symbol` would rule out IE](https://caniuse.com/?search=Symbol), unless a [`Symbol` polyfill](https://github.com/medikoo/es6-symbol) is applied. If this is a show-stopper, a new assumption `privateFieldsAsSymbolProperties` could be introduced that fixes the issue, but requires a target that supports `Symbol` (at least through a polyfill).

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by fwienber**
When babel-helpers are inlined, the global variable `id` was defined and initialized to zero in every generated file. This could lead to name clashes between #-private members in subclass and superclass, which leads to incorrect semantics.
The new solution uses `Symbol`s to create a unique private slot that cannot name-clash. I assume that in old environments that don't support `Symbol` natively, (babel-)polyfills must be loaded.

For details and discussion of other solutions, see #15389.

| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #15389
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | No
| Documentation PR Link    | https://github.com/babel/website/pull/2721
| Any Dependency Changes?  | No
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15393"></a>

### CODE REVIEW

**Comment by fwienber** [comment]
Next, I would like to add tests, but there are some specialties involved to create a reproducer (see my [codesandbox.io example](https://codesandbox.io/s/babel-privatefieldsasproperties-bug-gqvfsz?file=/package.json)):
* enable inlining of babel-helpers. This could probably be achieved by setting `"externalHelpers": false` in `options.json`.
* provide two input files, so that the helper code is inlined twice. Unfortunately, I don't see any way to achieve that with the current test setup.

**Comment by fwienber** [comment]
Since I force-pushed a solution using `globalThis` instead of evil `eval`, some tests under Node.js now fail with the error

    ReferenceError: globalThis is not defined

So obviously, `globalThis` is not available by default in Node.js. Where would I add the (trivial) polyfill for that?

**Comment by liuxingbaoyu** [comment]
It might be a bit early to work on this PR now. It's best to let us agree on which method to choose in the end first, to be able to use the best method and avoid wasting some time.

**Comment by fwienber** [comment]
Ok, so the test failures are a direct result from using `globalThis`. As expected, it must be polyfilled, but even in Node.js! Does anybody know where to add the (trivial) polyfill?

**Comment by fwienber** [comment]
> It might be a bit early to work on this PR now. It's best to let us agree on which method to choose in the end first, to be able to use the best method and avoid wasting some time. Anyway, thanks for your contribution!

This is why this PR is a "draft". I think a (draft) PR is the better place to discuss and try out concrete solutions, because it literally contains the code changes. I felt the issue was getting too long and into too much detail. But if you want, we can continue the discussion there.

**Comment by fwienber** [comment]
> Since I force-pushed a solution using `globalThis` instead of evil `eval`, some tests under Node.js now fail with the error
> 
> ```
> ReferenceError: globalThis is not defined
> ```
> 
> So obviously, `globalThis` is not available by default in Node.js. Where would I add the (trivial) polyfill for that?

It is a bit ironic that the "evil" solution passed all tests, but the more standard-compliant does not, isn't it?

**Comment by fwienber** [comment]
All tests are green, ready for review.

**Comment by fwienber** [comment]
@liuxingbaoyu what do you think? In the discussion in the corresponding issue, it seemed to me that everyone agrees that emitting `Symbol` (in contrast to using it directly in helper code) is fine.

**Review by JLHwung — APPROVED** [review]

File: packages/babel-helper-create-class-features-plugin/test/fixtures/plugin-proposal-private-methods/loose-true/output.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-var _privateMethod = /*#__PURE__*/babelHelpers.classPrivateFieldLooseKey("privateMethod");
+var _privateMethod = /*#__PURE__*/Symbol("privateMethod");
```
Comment (liuxingbaoyu) [inline packages/babel-helper-create-class-features-plugin/test/fixtures/plugin-proposal-private-methods/loose-true/output.js:1]: It looks like `/*#__PURE__*/` is no longer needed.
https://github.com/babel/babel/pull/15393/files#diff-328fd0199b81baa72df150181152575e627b62d082281c6b5ac8f0075a7174b6R88

**Review by liuxingbaoyu — APPROVED** [review]
My take on this PR is that we can try it out (see if anyone reports this after it's released, as it's technically a breaking change)

Even though there are two approvals now, if you want to review it! @nicolo-ribaudo

**Comment by nicolo-ribaudo** [comment]
I'm on the fence with this. In theory it should be good, but technically it's a breaking change. Maybe we could introduce this under a new `privateFieldsAsSymbols` assumption, and leave the existing behavior intact. We can then deprecate the old assumption documenting the problem, and force people to switch in Babel 8.

We will release a new minor roughly at the end of the week, and we could inclue the new assumption in that version.

**Comment by liuxingbaoyu** [comment]
Will the new assumptions be enabled by default? I'm afraid a lot of people haven't noticed it.

**Comment by nicolo-ribaudo** [comment]
Well, not even the old `privateFieldsAsProperties` was enabled by default. A possible timeline is:
- 7.21.0 - introduce the new assumption and document the problem in `privateFieldsAsProperties`
- 7.22.0 - if `@babel/plugin-transform-runtime` is not enabled, warn that `privateFieldsAsProperties` is deprecated
- 8.0.0 - remove `privateFieldsAsProperties`

**Comment by fwienber** [comment]
Sounds good!
So how do we proceed with this? I could try to make the corresponding code changes and adapt the PR. Or is anyone else going to take over? I don't want to retreat, just want to avoid duplicate work.

**Comment by liuxingbaoyu** [comment]
Feel free to do it!

**Comment by nicolo-ribaudo** [comment]
Feel free to make the changes here! If you need an example of how to define a new assumption option, you can check https://github.com/babel/babel/commit/7e50ee2d823ebc9e50eb3575beb77666214edf8e.

**Comment by nicolo-ribaudo** [comment]
I can work on docs for this PR.

**Comment by fwienber** [comment]
> Well, not even the old `privateFieldsAsProperties` was enabled by default. A possible timeline is:
> 
> * 7.21.0 - introduce the new assumption and document the problem in `privateFieldsAsProperties`
> * 7.22.0 - if `@babel/plugin-transform-runtime` is not enabled, warn that `privateFieldsAsProperties` is deprecated
> * 8.0.0 - remove `privateFieldsAsProperties`

What about the "loose" property of `plugin-proposal-class-properties`? If I get it right, it currently enables `privateFieldsAsProperties`. I am aware that ["assumptions" are now the preferred way to do this](https://babeljs.io/docs/en/babel-plugin-proposal-class-properties#loose), so would the behavior of using "loose" stay as-is in Babel 7 and change to enable `privateFieldsAsSymbols` instead in Babel 8?

**Comment by nicolo-ribaudo** [comment]
Yes. Or we might even just remove the `loose` option in Babel 8.

**Comment by fwienber** [comment]
More questions: If in 7.21.0/7.220, both `privateFieldsAsProperties` _and_ `privateFieldsAsSymbols` are set, which one should win? Should there be a warning in that case that the other assumption was ineffective?

File: packages/babel-helper-create-class-features-plugin/test/fixtures/plugin-proposal-private-methods/loose-true/output.js (hunk 1 above)
Comment (fwienber) [reply packages/babel-helper-create-class-features-plugin/test/fixtures/plugin-proposal-private-methods/loose-true/output.js:1]: I'm just about to remove the PURE comment for the new `Symbol` output and wondered whether it is needed for the existing standard solution, which generates `new WeakMap()` also with a PURE annotation. Shouldn't optimizer tools like Uglify _know_ that creating a (built-in) `WeakMap` has no side-effects?

Comment (fwienber) [reply packages/babel-helper-create-class-features-plugin/test/fixtures/plugin-proposal-private-methods/loose-true/output.js:1]: I'm talking about output like this:
https://github.com/babel/babel/blob/main/packages/babel-helper-create-class-features-plugin/test/fixtures/plugin-proposal-private-methods/loose-false/output.js#L1

**Comment by fwienber** [comment]
Since the new approach to introduce a dedicated assumption for using Symbol properties is quite different, I started a fresh branch and created a new PR #15435. I copied all relevant information (e.g. link to documentation PR), I hope I didn't miss anything. Let's continue there.

**Pull request closed without merging**